cli_tools/gce_windows_upgrade/upgrader/validators.go

```diff
@@ -0,0 +1,180 @@
+
+	computeClient daisyCompute.Client
+)
+
+func (u *Upgrader) validateParams() error {
+	if u.derivedVars == nil {
+		u.derivedVars = &derivedVars{}
+	}
+
+	if err := validation.ValidateStringFlagNotEmpty(u.ClientID, ClientIDFlagKey); err != nil {
+		return err
+	}
+	if err := validateOSVersion(u.SourceOS, u.TargetOS); err != nil {
+		return err
+	}
+	if err := validateInstanceURI(u.InstanceURI, u.derivedVars); err != nil {
+		return err
+	}
+	if err := validateInstance(u.derivedVars, u.SourceOS); err != nil {
+		return err
+	}
+
+	if u.Timeout == "" {
+		u.Timeout = DefaultTimeout
+	}
+
+	// Prepare resource names with a random suffix
+	suffix := path.RandString(8)
+	u.machineImageBackupName = fmt.Sprintf("windows-upgrade-backup-%v", suffix)
+	u.osDiskSnapshotName = fmt.Sprintf("windows-upgrade-backup-os-%v", suffix)
+	u.newOSDiskName = fmt.Sprintf("windows-upgraded-os-%v", suffix)
+	u.installMediaDiskName = fmt.Sprintf("windows-install-media-%v", suffix)
+
+	// Update 'project' value for logging purpose
+	*u.ProjectPtr = u.project
+
+	return nil
+}
+
+func validateOSVersion(sourceOS, targetOS string) error {
+	if sourceOS == "" {
+		return daisy.Errf("Flag -source-os must be provided. Please choose a supported version from {%v}.", strings.Join(SupportedSourceOSVersions(), ", "))
+	}
+	if _, ok := supportedSourceOSVersions[sourceOS]; !ok {
+		return daisy.Errf("Flag -source-os value '%v' unsupported. Please choose a supported version from {%v}.", sourceOS, strings.Join(SupportedSourceOSVersions(), ", "))
+	}
+	if targetOS == "" {
+		return daisy.Errf("Flag -target-os must be provided. Please choose a supported version from {%v}.", strings.Join(SupportedTargetOSVersions(), ", "))
+	}
+	if _, ok := supportedTargetOSVersions[targetOS]; !ok {
+		return daisy.Errf("Flag -target-os value '%v' unsupported. Please choose a supported version from {%v}.", targetOS, strings.Join(SupportedTargetOSVersions(), ", "))
+	}
+	// We may chain several upgrades together in the future (for example, 2008r2->2012r2->2016->2019).
+	// For now, we only support one-step upgrade.
+	if expectedTargetOS, _ := supportedSourceOSVersions[sourceOS]; expectedTargetOS != targetOS {
+		return daisy.Errf("Can't upgrade from %v to %v. Can only upgrade to %v.", sourceOS, targetOS, expectedTargetOS)
+	}
+	return nil
+}
+
+func validateInstanceURI(instanceURI string, derivedVars *derivedVars) error {
+	if instanceURI == "" {
+		return daisy.Errf("Flag -instance must be provided")
+	}
+	m := daisy.NamedSubexp(instanceURLRgx, instanceURI)
+	if m == nil {
+		return daisy.Errf("Please provide the instance flag in the form of 'projects/<project>/zones/<zone>/instances/<instance>', not %s", instanceURI)
+	}
+	derivedVars.project = m["project"]
+	derivedVars.zone = m["zone"]
+	derivedVars.instanceName = m["instance"]
+	return nil
+}
+
+func validateInstance(derivedVars *derivedVars, sourceOS string) error {
```

Comment:
Another example of a 'validate' function that makes modifications

Reply:
as above